RecoTracker/LSTCore/interface/alpaka/LSTInputCollection.h

```diff
@@ -0,0 +1,19 @@
+#ifndef RecoTracker_LSTCore_interface_alpaka_LSTInputCollection_h
+#define RecoTracker_LSTCore_interface_alpaka_LSTInputCollection_h
+
+#include <alpaka/alpaka.hpp>
+
+#include "DataFormats/Portable/interface/alpaka/PortableCollection.h"
+#include "RecoTracker/LSTCore/interface/LSTInputSoA.h"
+#include "RecoTracker/LSTCore/interface/LSTInputHostCollection.h"
+#include "RecoTracker/LSTCore/interface/LSTInputDeviceCollection.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/CopyToHost.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE::lst {
+  using LSTInputCollection = std::conditional_t<std::is_same_v<Device, alpaka::DevCpu>,
+                                                ::lst::LSTInputHostCollection,
+                                                ::lst::LSTInputDeviceCollection<Device>>;
```

Comment:
I wonder if
```suggestion
  using LSTInputCollection = std::conditional_t<std::is_same_v<ALPAKA_ACCELERATOR_NAMESPACE::Device, alpaka::DevCpu>,
                                                ::lst::LSTInputHostCollection,
                                                ::lst::LSTInputDeviceCollection<ALPAKA_ACCELERATOR_NAMESPACE::Device>>;
```
would make any difference wrt. the auto-parsing failure

Reply:
I tried this, and it gets replaced by this error:
```
error: no member named 'Device' in namespace 'ALPAKA_ACCELERATOR_NAMESPACE'
```
So the issue is that it's not including `HeterogeneousCore/AlpakaInterface/interface/config.h` correctly.

Reply:
Ok, thanks for the test. The leading order problem is nevertheless the header parsing itself, we'd want ROOT to not do that. I'm afraid we need to investigate more deeply what ROOT ends up doing to cause the header parsing.

@pcanal

Reply:
Is there something "original/creative" in this file? I thought that non-persisted heterogeneous data handling via `edm::Event` is supposed to work.

Reply:
No, by eye this file and the `classes_def.xml` look fine (or, similar-enough to other cases that have not caused problems).

Ok, looking at the aforementioned [log](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a5bdf1/45418/runTheMatrixCUDA-results/29634.704_TTbar_14TeV+Run4D110_lstOnGPUIters01TrackingOnly/step3_TTbar_14TeV+Run4D110_lstOnGPUIters01TrackingOnly.log) again I noticed the error message was about
```
edm::Wrapper<PortableDeviceMultiCollection<alpaka::DevUniformCudaHipRt<alpaka::ApiCudaRt>,lst::InputHitsSoALayout<128,false>,lst::InputPixelHitsSoALayout<128,false>,lst::InputPixelSeedsSoALayout<128,false> > >
``` 

With `PortableDevice...` there should be one layer of `edm::DeviceProduct` between it and the `edm::Wrapper` (as is in the `classes_def.xml`), i.e.
```
edm::Wrapper<edm::DeviceProduct<PortableDeviceMultiCollection<alpaka::DevUniformCudaHipRt<alpaka::ApiCudaRt>,lst::InputHitsSoALayout<128,false>,lst::InputPixelHitsSoALayout<128,false>,lst::InputPixelSeedsSoALayout<128,false> > > >
```

Then I'd guess this line
https://github.com/cms-sw/cmssw/blob/650b6436fbaabe0c2f165014bb671710ce76fafa/RecoTracker/LST/plugins/alpaka/LSTProducer.cc#L69
should be
```
    device::EDGetToken<lst::LSTInputCollection> lstInputToken_;
```
Ideally this should have resulted a `ProductNotFound` exception instead.

Reply:
Thank you Matti! That fixed the header parsing issue. However, now I'm back to a more standard error:
```
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: edm::DeviceProduct<PortableDeviceMultiCollection<alpaka::DevUniformCudaHipRt<alpaka::ApiCudaRt>,lst::InputHitsSoALayout<128,false>,lst::InputPixelHitsSoALayout<128,false>,lst::InputPixelSeedsSoALayout<128,false> > >
Looking for module label: lstInputProducer
Looking for productInstanceName: 
```
One thing that is not clear to me is whether I need to instantiate the `CopyToDevice<LSTInputCollection>` struct somewhere or do something else so that the host product can be fetched as a device product. Is there some way to print the available products to see where the mismatch might be?

Reply:
> ```
> Exception Message:
> Principal::getByToken: Found zero products matching all criteria
> Looking for type: edm::DeviceProduct<PortableDeviceMultiCollection<alpaka::DevUniformCudaHipRt<alpaka::ApiCudaRt>,lst::InputHitsSoALayout<128,false>,lst::InputPixelHitsSoALayout<128,false>,lst::InputPixelSeedsSoALayout<128,false> > >
> Looking for module label: lstInputProducer
> Looking for productInstanceName: 
> ```

This is produced by the `LSTInputProducer`, right? In that case, in order to benefit from the implicit host-to-device copy, you have to make the `LSTInputProducer` to an Alpaka module. This behavior is kind-of implied in https://github.com/cms-sw/cmssw/blob/master/HeterogeneousCore/AlpakaCore/README.md#edproducer, although the point of non-Alpaka modules not taking part in the implicit copies is not explicitly mentioned.

Reply:
Ah okay, I didn't realize it also had to be an Alpaka module, but that makes sense. Thanks!

Reply:
An alternative to the `conditional_t` would be
```suggestion
  using LSTInputCollection = PortableMultiCollection<InputHitsSoA, InputPixelHitsSoA, InputPixelSeedsSoA>;
```
Then `RecoTracker/LSTCore/interface/LSTInputDeviceCollection.h` could be removed.

Actually some other places follow a pattern where the aforementioned type alias would be called `LSTInputDeviceCollection`.

(I'm not sure if it would need `PortableMultiCollection3` for the "dictionary compilation to succeed")

Reply:
Thank you Matti. This issue is fixed now. There's a segfault happening at a later point, but I'm looking into it.

Reply:
> There's a segfault happening at a later point, but I'm looking into it

Nevermind, I was just doing something wrong. I think this should be fine now